prboom2/src/gl_main.c

```diff
@@ -2328,20 +2328,26 @@ static void gld_DrawSprite(GLSprite *sprite)
 
 static void gld_AddHealthBar(mobj_t* thing, GLSprite *sprite)
 {
-  if (((thing->flags & (MF_COUNTKILL | MF_CORPSE)) == MF_COUNTKILL) && (thing->health > 0))
+  if ((thing->flags & MF_SHOOTABLE) && thing->info->spawnhealth > 0 &&
+      thing->health > 0)
   {
     GLHealthBar hbar;
     int health_percent = thing->health * 100 / thing->info->spawnhealth;
 
-    hbar.cm = -1;
-    if (health_percent <= health_bar_red)
-      hbar.cm = CR_RED;
-    else if (health_percent <= health_bar_yellow)
-      hbar.cm = CR_YELLOW;
-    else if (health_percent <= health_bar_green)
-      hbar.cm = CR_GREEN;
+    if (health_percent <= health_bar_red) {
+      hbar.r = 1;
+      hbar.g = 0;
+    } else if (health_percent <= health_bar_yellow) {
+      hbar.r = 1;
+      hbar.g = (double)(health_percent - health_bar_red) /
+               (health_bar_yellow - health_bar_red);
+    } else if (health_percent <= health_bar_green) {
+      /* FullSimplify[1-((percent-yellow)/(green-yellow))] */
+      hbar.r = (double)(health_bar_green - health_percent) /
+               (health_bar_green - health_bar_yellow);
+      hbar.g = 1;
+    }
```

Comment:
Passes uninitialized values when the mobj health is 100%.

Following the logic here, there should be an else case with r=0;g=1.

Ex:

```c
 else {
      hbar.r = 0;
      hbar.g = 1;
    }
```